src/mesh/replicated_mesh.C

```diff
@@ -1417,5 +1417,151 @@ dof_id_type ReplicatedMesh::n_active_elem () const
                                                  this->active_elements_end()));
 }
 
+std::map<Point, std::vector<std::vector<Point>>>
+ReplicatedMesh::get_boundary_points() const
+{
+  // find number of disconnected subdomains
+  std::vector<Point> interior_points;
+  std::vector<subdomain_id_type> subdomains(max_elem_id() + 1, Elem::invalid_subdomain_id);
+
+  subdomain_id_type subdomain_counter = 0;
+
+  // a stack for visiting elements
+  std::vector<Elem *> list;
+  list.reserve(n_elem());
+
+  dof_id_type visited = 0;
+  do
+  {
+    for (const auto & elem : active_element_ptr_range())
+      if (subdomains[elem->id()] == Elem::invalid_subdomain_id)
+      {
+        list.push_back(elem);
+        interior_points.push_back(elem->centroid());
+        subdomains[elem->id()] = subdomain_counter;
+        break;
+      }
+    while (list.size() > 0)
+    {
+      // pop up an element
+      Elem * elem = list.back(); list.pop_back(); ++visited;
+
+      for (auto s : elem->side_index_range())
+      {
+        Elem * neighbor = elem->neighbor_ptr(s);
+        if (neighbor != nullptr && subdomains[neighbor->id()] == Elem::invalid_subdomain_id)
+        {
+          list.push_back(neighbor);
+          subdomains[neighbor->id()] = subdomain_counter;
+        }
+      }
+    }
+    std::cout << std::endl;
+    subdomain_counter++;
+  }
+  while (visited != n_elem());
+  // now subdomain_counter stores the number of disconnected subdomains
+
+  std::vector<std::vector<std::vector<Point>>> boundary_points(subdomain_counter);
+
+  // following boundary visiting algorithm only works with 2D
+  if (mesh_dimension() != 2)
+    libmesh_error_msg("Error: get_boundary_points only works for 2D now");
+
+  // get all boundary sides
+  // use a comparison functor to avoid run-time randomness due to pointers
```

Comment:
Now I'm second-guessing my "it won't matter if the particular representative point changes from run to run" claim earlier.  Elem pointers are going to stay fixed within the lifetime of any one call to a const function, so you actually are worrying about reproducing the exact ordering of these results between one run and another?

Reply:
Yes.

Reply:
Okay.  Perhaps rather than using a map key of a representative element centroid, then, you want to use something like the lowest element id?  Centroids might change by epsilon whenever we change certain mesh generation algorithms, whereas we can try to keep initial id numbrering consistent and libMesh+MOOSE already allow you to disable later renumbering.

Reply:
Maybe just a `Elem *`?

Reply:
Using id will make the value invalid after the mesh is renumbered possibly?

Reply:
Hmm... that may be a concern.  Usually anything you could do that would trigger a renumbering would also change geometry enough to invalidate the previous output, though, in which case you'd have to call this again anyway.  I guess all_second_order() can trigger a renumbering despite not changing element geometry ... but it *would* insert new boundary points and so you wouldn't want to reuse get_boundary_points() output in that case either.  So I don't see any cases where we'd have a *practical* concern.